File: Medium/0875-Koko Eating Bananas/Binary Search.rb
# @param {Integer[]} piles
# @param {Integer} h
# @return {Integer}
def min_eating_speed(piles, h)
  lo = 1
  hi = piles.max

  while lo < hi
    k = (lo + hi) / 2
    hours = piles.map { |x| (x - 1) / k + 1 }.sum

    if hours > h
      lo = k + 1
    else
      hi = k
    end
  end

  lo
end

File: Problemset/0875-Koko Eating Bananas/Binary Search.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut lo = 1;
        let mut hi = *piles.iter().max().unwrap();

        while lo < hi {
            let k = (lo + hi) / 2;
            let hours = piles.iter().map(|&x| ((x - 1) / k + 1) as i64).sum::<i64>();

            if hours > h as i64 {
                lo = k + 1;
            } else {
                hi = k;
            }
        }

        lo
    }
}
